clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -469,6 +472,333 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                OffsetCommitRequest.PartitionData partition2Data = commitRequest.offsetData().get(partition2);
+                // 5 records from topic1 partition and 0 from topic 2 partition have been returned to the consumer and auto commit is enabled;
+                // therefore, the committed offset of topic1 partition and topic2 partition should be 5 and 0, respectively.
+                if (partition1Data.offset == 5 && partition2Data.offset == 0) {
+                    commitAsExpected.set(true);
+                    return true;
+                }
+                return false;
+            }
+        }, offsetCommitResponse(responseMap), coordinator);
+
+        consumer.subscribe(Collections.singletonList(topic2));
+        assertTrue(commitAsExpected.get());
+    }
+
+    /**
+     * "consumer" subscribes to "topic1" and "topic2" (each with only one partition) and has auto-commit disabled.
+     * "consumer" reads 5 records from the single partition of "topic1" from offset 0.
+     * "consumer" changes subscription to "topic2" only.
+     * Since auto-commit is disabled, no offset commit request should be sent.
+     */
+    @Test
+    public void testSubscriptionChangeAndOffsetCommitWithAutoCommitDisabled() {
+        String topic1 = "topic1";
+        final TopicPartition partition1 = new TopicPartition(topic1, 0);
+        String topic2 = "topic2";
+        final TopicPartition partition2 = new TopicPartition(topic2, 0);
+    
+        int sessionTimeoutMs = 30000;
+        int heartbeatIntervalMs = 3000;
+    
+        // adjust auto commit interval lower than heartbeat so we don't need to deal with
+        // a concurrent heartbeat request
+        int autoCommitIntervalMs = 1000;
+    
+        Time time = new MockTime();
+        MockClient client = new MockClient(time);
+        Cluster cluster = TestUtils.singletonCluster(topic1, 1);
+        Node node = cluster.nodes().get(0);
+        client.setNode(node);
+        Metadata metadata = new Metadata(0, Long.MAX_VALUE);
+        metadata.update(cluster, time.milliseconds());
+        PartitionAssignor assignor = new RoundRobinAssignor();
+    
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
+                sessionTimeoutMs, heartbeatIntervalMs, false, autoCommitIntervalMs);
+        consumer.subscribe(Arrays.asList(topic1, topic2), new ConsumerRebalanceListener() {
+            @Override
+            public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+    
+            }
+    
+            @Override
+            public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                // set initial position so we don't need a lookup
+                for (TopicPartition partition : partitions)
+                    consumer.seek(partition, 0);
+            }
+        });
+    
+        // lookup coordinator
```

Comment:
Similarly, everything up to the fetch (i.e. coordinator lookup, join group, and sync group) are pretty much the same in all of these methods. Maybe we turn it into a function (e.g. `prepareRebalance`).